### PR TITLE
Allow `use super::*;` glob imports

### LINKED ISSUE #5554: Allow `use super::*;` in test modules
## Current behavior:
Using `use super::*;` in a test module throws a warning in pedantic mode:
```
warning: usage of wildcard import
   --> src/wifi_scan/mod.rs:104:9
    |
104 |     use super::*;
    |         ^^^^^^^^ help: try: `super::{ALLOWED_SYNCHRONOUS_RETRIES, DEVICE_OR_RESOURCE_BUSY_EXIT_CODE, SYNCHRONOUS_RETRY_DELAY_SECS, WifiIPInterface}`
```

## Expected behavior: 
No warning when `use super::*;` is used in a test module.

## Reasoning:

Using a wildcard import in test modules to bring in everything from the (parent) module being tested seems to be a very common idiom in Rust. It's recommended in [The Book](https://doc.rust-lang.org/book/ch11-01-writing-tests.html#adding-custom-failure-messages), it's more common in [the standard library](https://github.com/rust-lang/rust-clippy/files/4565003/super_in_rust_std.txt) than explicitly-listed imports, with the same bring true in a sample of [commonly-used crates](https://github.com/rust-lang/rust-clippy/files/4565004/super_in_rust_crates.txt).

Warning on this seems counter-productive - forcing explicit imports introduces a lot of additional friction to writing tests. For the first attempt at writing tests you _could_ use a wildcard import and then quickfix it to explicitly import, but any future tests written in that module will still require hopping up to the `use` statement and manually plugging in everything needed from the parent module. That's not the end of the world of course, but I'm always wary of anything that introduces unnecessary friction to the act of writing tests.

I think it makes a lot of sense to allow `use super::*;` in tests, in the same way [prelude::* imports are allowed](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/wildcard_imports.rs#L87). Something along the lines of `if current_module is [cfg(test)] and current_module_name matches "test[s]" and import is "super::*", do not warn`.

I'm happy to put up a pull request to implement this, if it sounds like a reasonable idea. I can reason from the above code how to match on "super", but I'd need some guidance on how to find the name of the current module and/or how to check for `#[cfg(test)]`.

### LINKED ISSUE #5569: Configuration variable for wildcard_imports to always warn
Suggested on [this pull request](https://github.com/rust-lang/rust-clippy/pull/5564#issuecomment-623181949):

> Since `wildcard_imports` is starting to warn in less cases, should it have a configuration variable to always warn? That would also help make the [documentation](https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_imports) clearer about it not warning in all cases by default.

It would be helpful if a configuration variable is added to `wildcard_imports` that skips checking for:

- [importing in a macro](https://github.com/rust-lang/rust-clippy/blob/76ddac5e8e9f588fda939cf8ec116e5926419758/clippy_lints/src/wildcard_imports.rs#L84)
- [importing the prelude](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/wildcard_imports.rs#L87)
- [importing in tests](https://github.com/rust-lang/rust-clippy/pull/5564)

This might be a good first issue.

### PR DESCRIPTION
**Pull request by MrAwesome**
changelog: Allow super::* glob imports

fixes #5554 
fixes #5569

A first pass at #5554 - this allows all `use super::*` to pass, which may or may not be desirable. The original issue was around allowing test modules to import their entire parent modules - I'm happy to modify this to do that instead, may just need some guidance on how to implement that (I played around a bit with #[cfg(test)] but from what I can gather, clippy itself isn't in test mode when running, even if the code in question is being checked for the test target).

### CODE REVIEW

**Comment by flip1995** [comment]
You cannot check for the `cfg(test)` attribute, since `cfg`/`cfg_attr`s don't exist anymore in the HIR.

**Comment by dylni** [comment]
Since `wildcard_imports` is starting to warn in less cases, should it have a configuration option to always warn? That would also help make the [documentation](https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_imports) clearer about it not warning in all cases by default.

**Comment by flip1995** [comment]
We could add this, yes

**Comment by MrAwesome** [comment]
Updated to check for "test" in the parent module name. I ended up doing this inline for simplicity's sake, happy to set a field on WildcardImports if that seems better.

**Comment by flip1995** [comment]
> I ended up doing this inline for simplicity's sake, happy to set a field on WildcardImports if that seems better.

That was my first idea when I reviewed this PR too. But that doesn't work, if the super import isn't directly in the main module. And allowing `use super::*` in test functions seems reasonable too. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0a611006689f73e0c3f70fa056b105d9) for test case.

**Comment by flip1995** [comment]
Better than a bool would be to add a counter how many test modules deep the lint check is, so it doesn't start linting in test modules again, just because a submodule is also called `*test*`.

**Comment by MrAwesome** [comment]
Updated to check for "test" in parent module/function names. 

@dylni I also went ahead and added the flag for #5569 since I'm already warmed up on the code.

File: clippy_lints/src/wildcard_imports.rs [hunk 1]
```diff
@@ -73,18 +73,38 @@ declare_clippy_lint! {
     "lint `use _::*` statements"
 }
 
-declare_lint_pass!(WildcardImports => [ENUM_GLOB_USE, WILDCARD_IMPORTS]);
+#[derive(Default)]
+pub struct WildcardImports {
+    warn_on_all: bool,
+    is_test_module: bool,
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:79]: This should not be necessary with the field below.

File: clippy_lints/src/wildcard_imports.rs [hunk 2]
```diff
@@ -152,6 +172,19 @@ impl LateLintPass<'_, '_> for WildcardImports {
             }
         }
     }
+
+    fn check_item_post(&mut self, _: &LateContext<'_, '_>, _: &Item<'_>) {
+        if self.is_test_module {
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:177]:
```suggestion
        if is_test_module(item) {
```

File: clippy_lints/src/wildcard_imports.rs [hunk 3]
```diff
@@ -168,9 +201,6 @@ fn is_super_only_import(segments: &[PathSegment<'_>]) -> bool {
     segments.len() == 1 && segments[0].ident.as_str() == "super"
 }
 
-fn is_in_test_module(cx: &LateContext<'_, '_>, item: &Item<'_>) -> bool {
-    let parent = cx.tcx.hir().get_parent_node(item.hir_id);
-    let parent_item = cx.tcx.hir().expect_item(parent);
-    let parent_name = parent_item.ident.name.as_str();
-    parent_name.contains("test")
+fn is_test_module(item: &Item<'_>) -> bool {
+    item.ident.name.as_str().contains("test")
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:205]: You also have to check the `ItemKind` here.

**Comment by MrAwesome** [comment]
Okay, updated to remove that field and check the itemkind. I checked for either `fn` or `mod`, should I do just `mod`? 

Also - are there docs I should update as well?

**Comment by dylni** [comment]
@MrAwesome Thanks! It's great to see it will be available immediately with this PR.

**Comment by flip1995** [comment]
Yes please exclude `Fn` here, since `test` is a pretty common string in a method name, like `test_for_xyz`, IMO.

**Comment by MrAwesome** [comment]
Updated to use Mod only, reflected that in the tests, and updated the docs to reflect all the new functionality.

File: clippy_lints/src/wildcard_imports.rs [hunk 4]
```diff
@@ -73,18 +78,35 @@ declare_clippy_lint! {
...
+        if is_test_module_or_function(item) {
+            self.test_modules_deep += 1;
+        }
         if_chain! {
-            if !in_macro(item.span);
+            if self.warn_on_all || !in_macro(item.span);
             if let ItemKind::Use(use_path, UseKind::Glob) = &item.kind;
-            // don't lint prelude glob imports
-            if !use_path.segments.iter().last().map_or(false, |ps| ps.ident.as_str() == "prelude");
+            if self.warn_on_all || !self.check_exceptions(use_path.segments);
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:109]: duplicated

File: clippy_lints/src/wildcard_imports.rs [hunk 5]
```diff
@@ -73,18 +78,35 @@ declare_clippy_lint! {
...
+impl_lint_pass!(WildcardImports => [ENUM_GLOB_USE, WILDCARD_IMPORTS]);
 
 impl LateLintPass<'_, '_> for WildcardImports {
     fn check_item(&mut self, cx: &LateContext<'_, '_>, item: &Item<'_>) {
         if item.vis.node.is_pub() || item.vis.node.is_pub_restricted() {
             return;
         }
+        if is_test_module_or_function(item) {
+            self.test_modules_deep += 1;
+        }
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:105]: You have to move this to the top of the function.

File: clippy_lints/src/wildcard_imports.rs [hunk 6]
```diff
@@ -153,4 +174,34 @@ impl LateLintPass<'_, '_> for WildcardImports {
             }
         }
     }
+
+    fn check_item_post(&mut self, _: &LateContext<'_, '_>, item: &Item<'_>) {
+        if is_test_module_or_function(item) {
+            self.test_modules_deep -= 1;
```
Comment (flip1995) [inline clippy_lints/src/wildcard_imports.rs:180]: To be save you can use `saturating_{sub,add}` instead.

File: clippy_lints/src/wildcard_imports.rs (hunk 4 above)
Comment (MrAwesome) [reply clippy_lints/src/wildcard_imports.rs]: Ah, ok - wasn't sure if the in_macro check still needed to live above the itemkind check.

**Comment by MrAwesome** [comment]
This look correct? Most of my experience is with Mercurial so I want to make extra sure I don't cause some weird repo state.

**Comment by flip1995** [comment]
Perfect, thanks!

@bors r+

**Pull request merged**